r-data-manipulation.Rmd

```diff
@@ -0,0 +1,1140 @@
+```{r, error = TRUE}
+my_object <- 12
+my_objet
+```
+you've used the wrong case, 
+```{r, error = TRUE}
+My_object
+```
+or you've forgotten that you've used a separator
+```{r, error = TRUE}
+myobject
+```
+
+This error also often arises when you forgot quotes around strings.  For example, if we want all the dogs that are male, and try
+```{r, error = TRUE}
+dog_licenses %>% filter(animal_gender == M)
+```
+you get an error because R is looking for an object called `M` to compare to the values in the column called `animal_gender`.  What we really wanted to do was compare the values in `animal_gender` to the string `"M"`:
+```{r}
+dog_licenses %>% filter(animal_gender == "M")
+```
+
+### Exercise: `object not found`
+
+Fix these `object not found` errors. (Hint: the names of the objects or variables being created should give you a clue to the intent of the code)
+
+*
+    ```{r, error = TRUE}
+    dog_licenses %>% 
+      mutate(year_issued = lubridate::year(Liscenceissuedate))
+    ```
+
+* 
+    ```{r, error = TRUE}
+    dogs_named_bruno <- dog_licenses %>% 
+      filter(animal_name == BRUNO)
+    ```
+
+### Warnings and messages 
+
+There are two other kinds of alerts R can give: warnings and messages.  These can both appear in the console with the same color as an error, but they are informational as opposed to fatal.  
+
+Messages are purely informational, for example when you read data in with `read_csv()` you get a message that describes the columns and their data types as parsed by the function:
+```{r, eval = FALSE}
+sites <- read_csv("site.csv")
+```
+```{r, echo = FALSE}
+sites <- read_csv(here("data", "site.csv"))
+```
+
+Warnings generally alert you that something was slightly unexpected but that R recovered and gave you a result anyway.  Poorly formatted CSV files will often result in warnings from `read_csv()`:
+```{r}
+bad_csv <- "
+id, 
+1, 2
+2, 1
+3, 5
+"
+bad <- read_csv(bad_csv)
+```
+Here there was a missing column name.  `read_csv()` still returns an object but the warning alerts you that it made some assumption to get it, i.e. that it made up a column name:
+```{r}
+bad
+```
+
+Warnings don't **stop** you from proceeding, but they should alert you to question whether you **should be** proceeding.
+
+### What do I do when I get an `Error` I can't fix?
+
+* Check that the error is reproducible.  Restart R with a clean slate and re-run your code up to and including the code that gives the error. This is the R version of the classic tech advice to "turn it off, then turn it on again".  TODO: link to reproducibility chapter.
+
+* Try googling it: for example, googling "R unexpected string constant" lead me to the question ["Error: unexpected symbol/input/string constant/numeric constant/SPECIAL in my code" on StackOverflow][error-unexpected] which gives some great examples of ways this error might arise.
```

Comment:
Can we be search engine neutral and say "Try searching for it online" or similar instead? Or does that make us sound like dinosaurs?

(On a more personal note, I like duckduckgo, but I have already mentioned "rubber duck debugging" and "ducktyping" in my python lecture, if I add "duckduckgoing" instead of "searching online", questions will be asked.) =)

Reply:
Good suggestion...I also use duckduckgo!